src/Dialect/ONNX/ONNXOps/Math/MatMul.cpp

```diff
@@ -189,6 +189,100 @@ LogicalResult ONNXMatMulIntegerOp::inferShapes(
   return shapeHelper.computeShapeAndUpdateType(elementType);
 }
 
+LogicalResult ONNXMatMulIntegerOp::verify() {
+  ONNXMatMulIntegerOp::Adaptor operandAdaptor(*this);
+  if (!hasShapeAndRank(getOperation()))
+    return success();
+
+  // If AZeroPoint is [M] (M != 1), A must be [M, K]
+  // If AZeroPoint rank is > 1, A must have the same rank, e.g.
+  // - [D1, D2, ..., DN, M, 1] if A is [D1, D2, ..., DN, M, K]
+  Value A = operandAdaptor.getA();
+  Value aZeroPoint = this->getAZeroPoint();
+  if (!isNoneValue(aZeroPoint)) {
+    auto aType = dyn_cast<ShapedType>(A.getType());
+    auto aZeroPointType = dyn_cast<ShapedType>(aZeroPoint.getType());
+    uint64_t aRank = aType.getRank();
+    uint64_t aZeroPointRank = aZeroPointType.getRank();
+    ArrayRef<int64_t> aShape = aType.getShape();
+    ArrayRef<int64_t> aZeroPointShape = aZeroPointType.getShape();
+    // If AZeroPoint is [M] (M != 1), A must be [M, K]
+    if ((aZeroPointRank == 1) && (!aZeroPointType.isDynamicDim(0)) &&
+        (aZeroPointShape[0] != 1) && (aRank != 2))
+      return onnx_mlir::Diagnostic::emitOperandHasUnexpectedRankError(
+          *this->getOperation(), A, aRank, "2");
+    // If AZeroPoint's rank is > 1, it must be the same as A's rank.
+    if (aZeroPointRank > 1) {
+      // Same rank.
+      if (aZeroPointRank != aRank)
+        return onnx_mlir::Diagnostic::emitInputsMustHaveSameRankError(
+            *this->getOperation(), "A", aRank, "aZeroPoint", aZeroPointRank);
+      // Broadcasting at the last dimension.
+      for (uint64_t i = 0; i < aRank - 1; ++i) {
+        if (!aType.isDynamicDim(i) && !aZeroPointType.isDynamicDim(i) &&
+            (aShape[i] != aZeroPointShape[i])) {
+          return onnx_mlir::Diagnostic::emitDimensionsMustHaveSameValueError(
+              *this->getOperation(), "A", i, aShape[i], "aZeroPoint", i,
+              aZeroPointShape[i]);
+        }
+      }
+      uint64_t lastAxis = aRank - 1;
+      if (!aType.isDynamicDim(lastAxis) &&
```

Comment:
I think you can leave out the check `!aType.isDynamicDim(lastAxis)` because aZeroPointType shouldn't have a non-1 static dim size in the last axis regardless of whether the dim size of a's last dimension is dynamic